File: public/codes/4-Clustering/2-Hierarchical_Clustering/HC.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 15 21:57:19 2023

@author: cruz
"""
# =============================================================================
# Clustering Jerarquico
# =============================================================================

# =============================================================================
# --------------------Importando librerias--------------------
# =============================================================================
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================

dataset = pd.read_csv('Mall_Customers.csv')

# Variables independiente:Mayuscula por ser una matriz.
X = dataset.iloc[:,[3,4]].values

# Variable dependiente:minuscula por ser un vector.
# y = dataset.iloc[:,-1].values 

# =============================================================================
# --------------------Tratamiendo de NAs--------------------
# =============================================================================
# No hace falta
# =============================================================================
# --------------------Codificar datos categoricos--------------------
# =============================================================================
# No hace falta
# =============================================================================
# --------------------Dividiendo dataset en conjuntos--------------------
# --------------------de entrenamiento y conjunto de testings------------------
# =============================================================================
# No hace falta
# =============================================================================
# --------------------Escalado de variables--------------------
# =============================================================================
# No hace falta
# =============================================================================
# ----Utilizando Dendrogramas para encontrar el numero optimo de cluster--
# =============================================================================
import scipy.cluster.hierarchy as sch

dendrogram = sch.dendrogram(sch.linkage(X,method="ward"))

plt.title("Dendrograma")
plt.xlabel("Clientes")
plt.ylabel("Distancia Euclidia")
plt.show()

# =============================================================================
# ----Ajustando el clustering jerarquico a los datos----
# =============================================================================
from sklearn.cluster import AgglomerativeClustering

hc=AgglomerativeClustering(n_clusters=5,affinity="euclidean",linkage="ward")
y_pred=hc.fit_predict(X)



plt.scatter(X[y_pred == 0, 0], X[y_pred == 0, 1], s=50, c="red",label = "Cluster1")
plt.scatter(X[y_pred == 1, 0], X[y_pred == 1, 1], s=50, c="blue",label = "Cluster2")
plt.scatter(X[y_pred == 2, 0], X[y_pred == 2, 1], s=50, c="green",label = "Cluster3")
plt.scatter(X[y_pred == 3, 0], X[y_pred == 3, 1], s=50, c="cyan",label = "Cluster4")
plt.scatter(X[y_pred == 4, 0], X[y_pred == 4, 1], s=50, c="magenta",label = "Cluster5")
plt.title("Cluster de clientes")
plt.xlabel("Ingresos anuales en miles de $")
plt.ylabel("Puntuacion de gastos")
plt.legend()
plt.show()

File: public/codes/4-Clustering/2-Hierarchical_Clustering/HC.R
# =============================================================================
# Clustering Jerarquico
# =============================================================================
# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================

dataset = read.csv('Mall_Customers.csv')
X=dataset[,4:5]

# No se hace ninguna distincion entre variables independientes 
# y variables dependientes en R
# =============================================================================
# --------------------Tratamiendo de NAs--------------------
# =============================================================================
# No aplica
# =============================================================================
# --------------------Codificar datos categoricos--------------------
# =============================================================================
# No aplica
# =============================================================================
# --------------------Dividiendo dataset en conjuntos--------------------
# --------------------de entrenamiento y conjunto de testings------------
# =============================================================================
# No aplica
# =============================================================================
# --------------------Escalado de variables--------------------
# =============================================================================
# No aplica
# =============================================================================
# ----Utilizando Dendrogramas para encontrar el numero optimo de cluster--
# =============================================================================
dendrogram=hclust(dist(X,method = "euclidean"),
                  method = "ward.D")
plot(dendrogram,
     main = "Dendrograma",
     xlab = "clientes del centro comercial",
     ylab = "distancia euclidea")
# =============================================================================
# ----------Aplicando el metodo Kmeans para segmentar el dataset---------
# =============================================================================
# Con ayuda de la grafica anterior sabemos el el k = 5 es el optimo
# k_means=kmeans(X,5,iter.max = 300, nstart = 10)

# =============================================================================
# ----Ajustando el clustering jerarquico a los datos----
# =============================================================================

hc=hclust(dist(X,method = "euclidean"),
                  method = "ward.D")
y_hc = cutree(hc,k=5,)
# =============================================================================
# Visualizacion de los datos
# =============================================================================
# library(cluster)

clusplot(X, y_hc, lines = 0, shade = TRUE, 
         color = TRUE,
         labels = 4,
         plotchar = FALSE,
         span = TRUE,
         main = "Clustering de clientes",
         xlab = "Ingresos anuales",
         ylab = "Puntuacion")








